src/protocols/hii.rs

```diff
@@ -0,0 +1,103 @@
+//! HII (Human Interface Infrastructure) Code Definitions
+//!
+//! This module contains bindings and definitions copied from Section 33.3 of
+//! the UEFI spec. If a package or section is listed here, it is included in its
+//! entirety. Other sections exist that may not be here yet.
+
+//
+// 33.3.1 Package Lists and Package Headers
+//
+#[repr(C)]
+#[derive(Debug, Clone, Copy)]
+pub struct PackageHeader {
+  pub length: [u8; 3],
+  pub r#type: u8,
+}
+
+pub const PACKAGE_TYPE_ALL : u8 = 0x00;
+pub const PACKAGE_TYPE_GUID : u8 = 0x01;
+pub const PACKAGE_FORMS : u8 = 0x02;
+pub const PACKAGE_STRINGS : u8 = 0x04;
+pub const PACKAGE_FONTS : u8 = 0x05;
+pub const PACKAGE_IMAGES : u8 = 0x06;
+pub const PACKAGE_SIMPLE_FONTS : u8 = 0x07;
+pub const PACKAGE_DEVICE_PATH : u8 = 0x08;
+pub const PACKAGE_KEYBOARD_LAYOUT : u8 = 0x09;
+pub const PACKAGE_ANIMATIONS : u8 = 0x0A;
+pub const PACKAGE_END : u8 = 0xDF;
+pub const PACKAGE_TYPE_SYSTEM_BEGIN : u8 = 0xE0;
+pub const PACKAGE_TYPE_SYSTEM_END : u8 = 0xFF;
+
+#[repr(C)]
+#[derive(Debug, Clone, Copy)]
+pub struct PackageListHeader {
+  pub package_list_guid: crate::base::Guid,
+  pub package_length: u32,
+}
+
+//
+// 33.3.3 Font Package
+//
+#[repr(C)]
+#[derive(Debug)]
+pub struct FontPackageHdr {
+  pub header: PackageHeader,
+  pub hdr_size: u32,
+  pub glyph_block_offset: u32,
+  pub cell: GlyphInfo,
+  pub font_style: FontStyle,
+  pub wide_glyphs: [crate::base::Char16],
+}
+
+pub type HiiFontStyle = u32;
+pub const FONT_STYLE_NORMAL: HiiFontStyle = 0x00000000;
+pub const FONT_STYLE_BOLD: HiiFontStyle = 0x00000001;
+pub const FONT_STYLE_ITALIC: HiiFontStyle = 0x00000002;
+pub const FONT_STYLE_EMBOSS: HiiFontStyle = 0x00010000;
+pub const FONT_STYLE_OUTLINE: HiiFontStyle = 0x00020000;
+pub const FONT_STYLE_SHADOW: HiiFontStyle = 0x00040000;
+pub const FONT_STYLE_UNDERLINE: HiiFontStyle = 0x00080000;
+pub const FONT_STYLE_DBL_UNDER: HiiFontStyle = 0x00100000;
+
+#[repr(C)]
+#[derive(Debug)]
+pub struct GlyphBlock {
+  pub block_type: u8,
+  pub block_body: [u8],
+}
+
+pub const GIBT_END: u8 = 0x00;
+pub const GIBT_GLYPH: u8 = 0x10;
+pub const GIBT_GLYPHS: u8 = 0x11;
+pub const GIBT_GLYPH_DEFAULT: u8 = 0x12;
+pub const GIBT_GLYPHS_DEFAULT: u8 = 0x13;
+pub const GIBT_GLYPH_VARIABILITY: u8 = 0x14;
+pub const GIBT_DUPLICATE: u8 = 0x20;
+pub const GIBT_SKIP2: u8 = 0x21;
+pub const GIBT_SKIP1: u8 = 0x22;
+pub const GIBT_DEFAULTS: u8 = 0x23;
+pub const GIBT_EXT1: u8 = 0x30;
+pub const GIBT_EXT2: u8 = 0x31;
+pub const GIBT_EXT4: u8 = 0x32;
+
+#[repr(C)]
+#[derive(Debug, Clone, Copy)]
+pub struct GlyphInfo {
+  pub width: u16,
+  pub height: u16,
+  pub offset_x: i16,
+  pub offset_y: i16,
+  pub advance_x: i16,
+}
+
+#[repr(C)]
+#[derive(Debug, Clone)]
+pub struct GibtDefaultsBlock {
```

Comment:
So... then I got to here, where the spec says that it's two substructures as fields of a parent, but in reality is a dynamically-size serialization of two open-ended structures (EFI_HII_GLYPH_BLOCK and EFI_HII_GLYPH_INFO).

If we look at the Tianocore implementation (which is basically the industry standard), even they don't code it according to the spec:
https://github.com/tianocore/edk2/blob/8028b2907e20b21cd7d69639a36ac82a77c81dc1/MdePkg/Include/Uefi/UefiInternalFormRepresentation.h#L216

How have you been addressing things like this that are effectively impossible to transcribe?

Reply:
@dvdhrm

Reply:
I would follow what tianocore does, which in this case means dropping the `glyph_body` dynamically-sized member. I think I even did that before for the device-path structures. So yeah, I don't mind dropping those 0-size-tail-members.